etcd-launcher does not successfully recover from PVC deletion

### What happened?

This was originally raised as a PR (#9019).

When deleting a single PVC for etcd when `etcd-launcher` is enabled, the fresh replacement pod does not successfully start but instead fails.

```
022-02-16 12:53:27.829464 I | etcdserver: starting server... [version: 3.4.3, cluster version: to_be_decided]
2022-02-16 12:53:27.832051 I | embed: ClientTLS: cert = /etc/etcd/pki/tls/etcd-tls.crt, key = /etc/etcd/pki/tls/etcd-tls.key, trusted-ca = /etc/etcd/pki/ca/ca.crt, client-cert-auth = true, crl-file =
2022-02-16 12:53:27.832081 I | embed: listening for peers on 10.44.5.194:2381
2022-02-16 12:53:27.832289 I | embed: listening for metrics on http://127.0.0.1:2378
2022-02-16 12:53:27.832428 I | embed: listening for metrics on http://10.44.5.194:2378
2022-02-16 12:53:27.859967 I | rafthttp: started HTTP pipelining with peer f06ca2eeaf5c8ad8
raft2022/02/16 12:53:27 INFO: a4adad3686f23568 [term: 0] received a MsgHeartbeat message with higher term from f06ca2eeaf5c8ad8 [term: 8]
raft2022/02/16 12:53:27 INFO: a4adad3686f23568 became follower at term 8
raft2022/02/16 12:53:27 tocommit(2292) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
panic: tocommit(2292) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

goroutine 127 [running]:
log.(*Logger).Panicf(0xc0000c4aa0, 0x10c8397, 0x5d, 0xc000217c40, 0x2, 0x2)
	/usr/local/go/src/log/log.go:219 +0xc1
go.etcd.io/etcd/raft.(*DefaultLogger).Panicf(0x1aa0040, 0x10c8397, 0x5d, 0xc000217c40, 0x2, 0x2)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/logger.go:127 +0x60
go.etcd.io/etcd/raft.(*raftLog).commitTo(0xc000219c00, 0x8f4)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/log.go:203 +0x131
go.etcd.io/etcd/raft.(*raft).handleHeartbeat(0xc000183680, 0x8, 0xa4adad3686f23568, 0xf06ca2eeaf5c8ad8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/raft.go:1396 +0x54
go.etcd.io/etcd/raft.stepFollower(0xc000183680, 0x8, 0xa4adad3686f23568, 0xf06ca2eeaf5c8ad8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/raft.go:1341 +0x480
go.etcd.io/etcd/raft.(*raft).Step(0xc000183680, 0x8, 0xa4adad3686f23568, 0xf06ca2eeaf5c8ad8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/raft.go:984 +0x1113
go.etcd.io/etcd/raft.(*node).run(0xc0002d8420)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/node.go:352 +0xab6
created by go.etcd.io/etcd/raft.RestartNode
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/node.go:240 +0x33c
{"level":"info","time":"2022-02-16T12:53:28.840Z","caller":"etcd-launcher/main.go:139","msg":"[https://etcd-1.etcd.cluster-g4mp4z6w52.svc.cluster.local:2381] is a member"}
{"level":"panic","time":"2022-02-16T12:53:28.840Z","caller":"etcd-launcher/main.go:174","msg":"exit status 2"}
panic: exit status 2

goroutine 1 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000036780, {0x0, 0x0, 0x0})
	go.uber.org/zap@v1.19.1/zapcore/entry.go:232 +0x446
go.uber.org/zap.(*SugaredLogger).log(0xc0000b77e8, 0x4, {0x0, 0xe}, {0xc00098faf0, 0x1, 0x1}, {0x0, 0x0, 0x0})
	go.uber.org/zap@v1.19.1/sugar.go:227 +0xee
go.uber.org/zap.(*SugaredLogger).Panic(...)
	go.uber.org/zap@v1.19.1/sugar.go:123
main.main()
	k8c.io/kubermatic/v2/cmd/etcd-launcher/main.go:174 +0x1396
```

### What should happen?

`etcd-launcher` is successfully adding the fresh pod as new node to recover from data loss.

### How to reproduce?

1. Create cluster.
2. Enable feature gate `etcdLauncher` on cluster.
3. Force deletion of a PV attached to an etcd PVC (delete the PVC and delete matching Pod).
4. See that Pod is failing.

### Environment

I tested this on the following environment. @furkhat probably used a different version.

- KKP Version: `master`
- Shared or separate master/seed clusters?: shared
- Master/seed cluster provider: -

--- a/cmd/etcd-launcher/main.go
+++ b/cmd/etcd-launcher/main.go
@@ -124,6 +124,35 @@ func main() {
 		}
 	}
 
+	thisMember, err := e.getMemberByName(e.podName, log)
+	if err != nil {
+		log.Warnw("failed to check cluster membership", zap.Error(err))
+	}
+
+	if thisMember != nil {
+		log.Infof("%v is a member", thisMember.GetPeerURLs())
+
+		if _, err := os.Stat(filepath.Join(e.dataDir, "member")); errors.Is(err, fs.ErrNotExist) {
+			client, err := e.getClusterClient()
+			if err != nil {
+				log.Panicw("can't find cluster client: %v", zap.Error(err))
+			}
+
+			log.Warnw("No data dir, removing stale membership to rejoin cluster as new member")
+
+			_, err = client.MemberRemove(context.Background(), thisMember.ID)
+			if err != nil {
+				closeClient(client, log)
+				log.Panicw("remove itself due to data dir loss", zap.Error(err))
+			}
+
+			closeClient(client, log)
+			if err := joinCluster(e, log); err != nil {
+				log.Panicw("join cluster as fresh member", zap.Error(err))
+			}
+		}
+	}
+
 	// setup and start etcd command
 	etcdCmd, err := startEtcdCmd(e, log)
 	if err != nil {
@@ -134,33 +163,6 @@ func main() {
 		return e.isClusterHealthy(log)
 	}); err != nil {
 		log.Panicw("manager thread failed to connect to cluster", zap.Error(err))
-	}
-
-	thisMember, err := e.getMemberByName(e.podName, log)
-	if err != nil {
-		log.Panicw("failed to check cluster membership", zap.Error(err))
-	}
-	if thisMember != nil {
-		log.Infof("%v is a member", thisMember.GetPeerURLs())
-
-		if _, err := os.Stat(filepath.Join(e.dataDir, "member")); errors.Is(err, fs.ErrNotExist) {
-			client, err := e.getClusterClient()
-			if err != nil {
-				log.Panicw("can't find cluster client: %v", zap.Error(err))
-			}
-			log.Warnw("No data dir, to ensure recovery removing and adding the member")
-			_, err = client.MemberRemove(context.Background(), thisMember.ID)
-			if err != nil {
-				closeClient(client, log)
-				log.Panicw("remove itself due to data dir loss", zap.Error(err))
-			}
-			closeClient(client, log)
-			if err := joinCluster(e, log); err != nil {
-				log.Panicw("join cluster as fresh member", zap.Error(err))
-			}
-		}
-	} else if err := joinCluster(e, log); err != nil {
-		log.Panicw("join cluster as fresh member", zap.Error(err))
 	}
 
 	// reconcile dead members continuously. Initially we did this once as a step at the end of start up. We did that because scale up/down operations required a full restart of the ring with each node add/remove. However, this is no longer the case, so we need to separate the reconcile from the start up process and do it continuously.

--- a/cmd/etcd-launcher/main.go
+++ b/cmd/etcd-launcher/main.go
@@ -125,9 +125,11 @@ func main() {
 	}
 
 	thisMember, err := e.getMemberByName(e.podName, log)
-	if err != nil {
+
+	switch {
+	case err != nil:
 		log.Warnw("failed to check cluster membership", zap.Error(err))
-	} else if thisMember != nil {
+	case thisMember != nil:
 		log.Infof("%v is a member", thisMember.GetPeerURLs())
 
 		if _, err := os.Stat(filepath.Join(e.dataDir, "member")); errors.Is(err, fs.ErrNotExist) {
@@ -149,7 +151,7 @@ func main() {
 				log.Panicw("join cluster as fresh member", zap.Error(err))
 			}
 		}
-	} else {
+	default:
 		// if no membership information was found but we were able to list from an etcd cluster, we can attempt to join
 		if err := joinCluster(e, log); err != nil {
 			log.Panicw("failed to join cluster as fresh member", zap.Error(err))

--- a/cmd/etcd-launcher/main.go
+++ b/cmd/etcd-launcher/main.go
@@ -127,9 +127,7 @@ func main() {
 	thisMember, err := e.getMemberByName(e.podName, log)
 	if err != nil {
 		log.Warnw("failed to check cluster membership", zap.Error(err))
-	}
-
-	if thisMember != nil {
+	} else if thisMember != nil {
 		log.Infof("%v is a member", thisMember.GetPeerURLs())
 
 		if _, err := os.Stat(filepath.Join(e.dataDir, "member")); errors.Is(err, fs.ErrNotExist) {
@@ -150,6 +148,11 @@ func main() {
 			if err := joinCluster(e, log); err != nil {
 				log.Panicw("join cluster as fresh member", zap.Error(err))
 			}
+		}
+	} else {
+		// if no membership information was found but we were able to list from an etcd cluster, we can attempt to join
+		if err := joinCluster(e, log); err != nil {
+			log.Panicw("failed to join cluster as fresh member", zap.Error(err))
 		}
 	}
 

--- a/pkg/test/e2e/etcd-launcher/etcd_test.go
+++ b/pkg/test/e2e/etcd-launcher/etcd_test.go
@@ -211,16 +211,71 @@ func TestScaling(t *testing.T) {
 	}
 	waitForQuorum(t)
 
-	if err := breakAndRecover(ctx, t, client, cluster); err != nil {
-		t.Fatalf("failed to test volume recovery: %v", err)
-	}
-	waitForQuorum(t)
-
 	if err := disableLauncher(ctx, t, client, cluster); err != nil {
 		t.Fatalf("succeeded in disabling immutable feature etcd-launcher: %v", err)
 	}
 
 	t.Log("tests succeeded")
+}
+
+func TestRecovery(t *testing.T) {
+	ctx := context.Background()
+
+	client, _, _, err := utils.GetClients()
+	if err != nil {
+		t.Fatalf("failed to get client for seed cluster: %v", err)
+	}
+
+	// login
+	masterToken, err := utils.RetrieveMasterToken(ctx)
+	if err != nil {
+		t.Fatalf("failed to get master token: %v", err)
+	}
+	testClient := utils.NewTestClient(masterToken, t)
+
+	// create dummy project
+	t.Log("creating project...")
+	project, err := testClient.CreateProject(rand.String(10))
+	if err != nil {
+		t.Fatalf("failed to create project: %v", err)
+	}
+	defer cleanupProject(t, project.ID)
+
+	// create dummy cluster (NB: If these tests fail, the etcd ring can be
+	// _so_ dead that any cleanup attempt is futile; make sure to not create
+	// any cloud resources, as they might be orphaned)
+
+	t.Log("creating cluster...")
+	apiCluster, err := testClient.CreateHetznerCluster(project.ID, datacenter, rand.String(10), credential, version, location, 0)
+	if err != nil {
+		t.Fatalf("failed to create cluster: %v", err)
+	}
+
+	// wait for the cluster to become healthy
+	if err := testClient.WaitForClusterHealthy(project.ID, datacenter, apiCluster.ID); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
+
+	// get the cluster object (the CRD, not the API's representation)
+	cluster := &kubermaticv1.Cluster{}
+	if err := client.Get(ctx, types.NamespacedName{Name: apiCluster.ID}, cluster); err != nil {
+		t.Fatalf("failed to get cluster: %v", err)
+	}
+
+	if err := enableLauncher(ctx, t, client, cluster); err != nil {
+		t.Fatalf("failed to enable etcd-launcher: %v", err)
+	}
+	waitForQuorum(t)
+
+	if err := breakAndRecoverPV(ctx, t, client, cluster); err != nil {
+		t.Fatalf("failed to test volume recovery: %v", err)
+	}
+	waitForQuorum(t)
+
+	if err := breakAndRecoverPVC(ctx, t, client, cluster); err != nil {
+		t.Fatalf("failed to recover from PVC deletion: %v", err)
+	}
+	waitForQuorum(t)
 }
 
 func createBackup(ctx context.Context, t *testing.T, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) (error, *kubermaticv1.EtcdBackupConfig) {
@@ -353,7 +408,7 @@ func scaleDown(ctx context.Context, t *testing.T, client ctrlruntimeclient.Clien
 	return nil
 }
 
-func breakAndRecover(ctx context.Context, t *testing.T, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) error {
+func breakAndRecoverPV(ctx context.Context, t *testing.T, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) error {
 	// delete one of the etcd node PVs
 	t.Log("testing etcd node PV automatic recovery...")
 	if err := forceDeleteEtcdPV(ctx, client, cluster); err != nil {
@@ -369,6 +424,24 @@ func breakAndRecover(ctx context.Context, t *testing.T, client ctrlruntimeclient
 		return fmt.Errorf("etcd cluster is not healthy: %w", err)
 	}
 	t.Log("etcd node PV recovered successfully.")
+
+	return nil
+}
+
+func breakAndRecoverPVC(ctx context.Context, t *testing.T, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) error {
+	// delete one of the etcd node PVCs
+	t.Log("testing etcd-launcher recovery from deleted PVC ...")
+	if err := deleteEtcdPVC(ctx, client, cluster); err != nil {
+		return fmt.Errorf("failed to delete etcd node PVC: %w", err)
+	}
+
+	time.Sleep(30 * time.Second)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		return fmt.Errorf("etcd cluster is not healthy: %w", err)
+	}
+
+	t.Log("etcd node recovered from PVC deletion successfully.")
 
 	return nil
 }
@@ -631,6 +704,57 @@ func forceDeleteEtcdPV(ctx context.Context, client ctrlruntimeclient.Client, clu
 	return wait.PollImmediate(2*time.Second, 3*time.Minute, func() (bool, error) {
 		if err := client.Get(ctx, typedName, pv); kerrors.IsNotFound(err) {
 			return true, nil
+		}
+		return false, nil
+	})
+}
+
+func deleteEtcdPVC(ctx context.Context, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) error {
+	ns := clusterNamespace(cluster)
+
+	selector, err := labels.Parse("app=etcd")
+	if err != nil {
+		return fmt.Errorf("failed to parse label selector: %w", err)
+	}
+
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	opt := &ctrlruntimeclient.ListOptions{
+		LabelSelector: selector,
+		Namespace:     ns,
+	}
+	if err := client.List(ctx, pvcList, opt); err != nil || len(pvcList.Items) == 0 {
+		return fmt.Errorf("failed to list PVCs or empty list in cluster namespace: %w", err)
+	}
+
+	// pick a random PVC and get the corresponding pod
+	index := rand.Intn(len(pvcList.Items))
+	pvc := pvcList.Items[index]
+	oldPvc := pvc.DeepCopy()
+
+	podList := &corev1.PodList{}
+	if err := client.List(ctx, podList, opt); err != nil || len(podList.Items) != len(pvcList.Items) {
+		return fmt.Errorf("failed to list etcd pods or bad number of pods: %w", err)
+	}
+
+	pod := podList.Items[index]
+
+	// first, we delete it
+	if err := client.Delete(ctx, &pvc); err != nil {
+		return fmt.Errorf("failed to delete etcd node PVC %s: %w", pvc.Name, err)
+	}
+
+	// now, we delete the pod so the PVC can be finalised
+	if err := client.Delete(ctx, &pod); err != nil {
+		return fmt.Errorf("failed to delete etcd pod %s: %w", pod.Name, err)
+	}
+
+	// make sure the PVC is recreated by checking the CreationTimestamp against a DeepCopy
+	// created of the PVC resource.
+	return wait.PollImmediate(2*time.Second, 3*time.Minute, func() (bool, error) {
+		if err := client.Get(ctx, types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}, &pvc); err == nil {
+			if oldPvc.ObjectMeta.CreationTimestamp.Before(&pvc.ObjectMeta.CreationTimestamp) {
+				return true, nil
+			}
 		}
 		return false, nil
 	})

--- a/pkg/test/e2e/etcd-launcher/etcd_test.go
+++ b/pkg/test/e2e/etcd-launcher/etcd_test.go
@@ -613,10 +613,10 @@ func waitForEtcdRestore(ctx context.Context, t *testing.T, client ctrlruntimecli
 }
 
 func waitForClusterHealthy(ctx context.Context, t *testing.T, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) error {
+	before := time.Now()
+
 	// let's briefly sleep to give controllers a chance to kick in
 	time.Sleep(10 * time.Second)
-
-	before := time.Now()
 
 	if err := wait.PollImmediate(3*time.Second, 10*time.Minute, func() (bool, error) {
 		// refresh cluster object for updated health status

--- a/pkg/test/e2e/etcd-launcher/etcd_test.go
+++ b/pkg/test/e2e/etcd-launcher/etcd_test.go
@@ -130,7 +130,10 @@ func TestBackup(t *testing.T) {
 	if err := enableLauncher(ctx, t, client, cluster); err != nil {
 		t.Fatalf("failed to enable etcd-launcher: %v", err)
 	}
-	waitForQuorum(t)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 
 	// restore from backup
 	if err := restoreBackup(ctx, t, client, cluster, backup); err != nil {
@@ -138,7 +141,9 @@ func TestBackup(t *testing.T) {
 	}
 	t.Log("restored etcd backup")
 
-	waitForQuorum(t)
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 
 	// check if resource was restored
 	restoredNamespace := &corev1.Namespace{}
@@ -199,17 +204,26 @@ func TestScaling(t *testing.T) {
 	if err := enableLauncher(ctx, t, client, cluster); err != nil {
 		t.Fatalf("failed to enable etcd-launcher: %v", err)
 	}
-	waitForQuorum(t)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 
 	if err := scaleUp(ctx, t, client, cluster); err != nil {
 		t.Fatalf("failed to scale up: %v", err)
 	}
-	waitForQuorum(t)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 
 	if err := scaleDown(ctx, t, client, cluster); err != nil {
 		t.Fatalf("failed to scale down: %v", err)
 	}
-	waitForQuorum(t)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 
 	if err := disableLauncher(ctx, t, client, cluster); err != nil {
 		t.Fatalf("succeeded in disabling immutable feature etcd-launcher: %v", err)
@@ -265,17 +279,26 @@ func TestRecovery(t *testing.T) {
 	if err := enableLauncher(ctx, t, client, cluster); err != nil {
 		t.Fatalf("failed to enable etcd-launcher: %v", err)
 	}
-	waitForQuorum(t)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 
 	if err := breakAndRecoverPV(ctx, t, client, cluster); err != nil {
 		t.Fatalf("failed to test volume recovery: %v", err)
 	}
-	waitForQuorum(t)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 
 	if err := breakAndRecoverPVC(ctx, t, client, cluster); err != nil {
 		t.Fatalf("failed to recover from PVC deletion: %v", err)
 	}
-	waitForQuorum(t)
+
+	if err := waitForClusterHealthy(ctx, t, client, cluster); err != nil {
+		t.Fatalf("cluster did not become healthy: %v", err)
+	}
 }
 
 func createBackup(ctx context.Context, t *testing.T, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) (error, *kubermaticv1.EtcdBackupConfig) {
@@ -590,6 +613,9 @@ func waitForEtcdRestore(ctx context.Context, t *testing.T, client ctrlruntimecli
 }
 
 func waitForClusterHealthy(ctx context.Context, t *testing.T, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) error {
+	// let's briefly sleep to give controllers a chance to kick in
+	time.Sleep(10 * time.Second)
+
 	before := time.Now()
 
 	if err := wait.PollImmediate(3*time.Second, 10*time.Minute, func() (bool, error) {
@@ -653,12 +679,6 @@ func waitForRollout(ctx context.Context, t *testing.T, client ctrlruntimeclient.
 	}
 
 	return nil
-}
-
-// TODO: Make this much smarter.
-func waitForQuorum(t *testing.T) {
-	t.Log("waiting for etcd to regain quorum...")
-	time.Sleep(2 * time.Minute)
 }
 
 func forceDeleteEtcdPV(ctx context.Context, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster) error {

